subsys/net/l2/virtual/ipip/ipip.c

```diff
@@ -0,0 +1,575 @@
+/*
+ * Copyright (c) 2021 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_virtual_ipip, CONFIG_NET_L2_IPIP_LOG_LEVEL);
+
+#include <zephyr.h>
+#include <device.h>
+#include <errno.h>
+
+#include <net/net_core.h>
+#include <net/net_ip.h>
+#include <net/virtual.h>
+
+#include "ipv4.h"
+#include "ipv6.h"
+#include "net_private.h"
+
+#if defined(CONFIG_NET_L2_IPIP_TXRX_DEBUG)
+#define DEBUG_TX 1
+#define DEBUG_RX 1
+#else
+#define DEBUG_TX 0
+#define DEBUG_RX 0
+#endif
+
+#define IPIPV4_MTU NET_IPV4_MTU
+#define IPIPV6_MTU NET_IPV6_MTU
+
+#define PKT_ALLOC_TIME K_MSEC(50)
+
+struct ipip_context {
+	struct net_if *iface;
+	struct net_if *attached_to;
+	union {
+		sa_family_t family;
+		struct net_addr peer;
+	};
+
+	union {
+		const struct in_addr *my4addr;
+		const struct in6_addr *my6addr;
+	};
+
+	bool status;
+	bool init_done;
+};
+
+static int ipip_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 0;
+}
+
+static void iface_init(struct net_if *iface)
+{
+	struct ipip_context *ctx = net_if_get_device(iface)->data;
+
+	if (ctx->init_done) {
+		return;
+	}
+
+	ctx->iface = iface;
+	net_if_flag_set(iface, NET_IF_NO_AUTO_START);
+	net_if_flag_set(iface, NET_IF_POINTOPOINT);
+
+	(void)net_virtual_set_flags(iface, NET_L2_POINT_TO_POINT);
+
+	ctx->init_done = true;
+}
+
+static enum virtual_interface_caps get_capabilities(struct net_if *iface)
+{
+	ARG_UNUSED(iface);
+
+	return (enum virtual_interface_caps)0;
+}
+
+static int interface_start(const struct device *dev)
+{
+	struct ipip_context *ctx = dev->data;
+	int ret = 0;
+
+	if (ctx->status) {
+		return -EALREADY;
+	}
+
+	ctx->status = true;
+
+	NET_DBG("Starting iface %d", net_if_get_by_iface(ctx->iface));
+
+	/* You can implement here any special action that is needed
```

Comment:
Should this comment be here? I'd expect it more in a sample code than an actual tunneling driver.

Reply:
> Should this comment be here? I'd expect it more in a sample code than an actual tunneling driver.

Yeah, a left over from some other ideas. I will remove this.